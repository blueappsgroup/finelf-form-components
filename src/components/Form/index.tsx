import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
