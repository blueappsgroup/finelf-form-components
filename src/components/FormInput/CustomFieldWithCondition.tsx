import React, { FC, useContext, useEffect, useState } from 'react'
import { Field, useFormikContext } from 'formik'
import { FormContext, setFormValuesToCache } from '../../utils'

type Props = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  requiredCondition?: { [key: string]: any }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  visibleCondition?: { [key: string]: any }
}
const CustomFieldWithCondition: FC<Props> = (props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const { values, errors, setErrors, setValues } = useFormikContext<any>()
  const { id } = useContext(FormContext)
  const [required, setRequired] = useState(props.required || false)
  const [visible, setVisible] = useState<boolean>(
    (props.visibleCondition && false) || true
  )

  useEffect(() => {
    if (!props.required && props.requiredCondition) {
      let isRequired = false

      Object.keys(props.requiredCondition).some((key) => {
        if (
          props.requiredCondition &&
          props.requiredCondition[key].includes(values[key])
        ) {
          isRequired = true
          return true
        }

        return false
      })

      if (required !== isRequired) {
        setRequired(isRequired)
      }
    }
  }, [values, props.required, props.requiredCondition, required, props.name])

  useEffect(() => {
    if (props.visibleCondition) {
      let isVisible = false

      Object.keys(props.visibleCondition).some((key) => {
        if (
          props.visibleCondition &&
          props.visibleCondition[key].includes(values[key])
        ) {
          isVisible = true

          return true
        }

        return false
      })

      if (!isVisible && isVisible !== visible) {
        setValues({ ...values, [props.name]: undefined }, false)
        setErrors({ ...errors, [props.name]: undefined })
        setFormValuesToCache({ ...values, [props.name]: undefined }, id)
        setVisible(isVisible)
      }

      if (isVisible && isVisible !== visible) {
        setVisible(isVisible)
      }
    }
  }, [
    values,
    setValues,
    props.visibleCondition,
    props.name,
    visible,
    errors,
    setErrors,
    id,
  ])

  return (
    <Field
      {...props}
      required={required && visible}
      validate={props.validate(required && visible)}
      visible={visible}
    />
  )
}

export default CustomFieldWithCondition
