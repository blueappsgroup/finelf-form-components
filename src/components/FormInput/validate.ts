import {
  isEmpty,
  isNotValidLength,
  isNotValidPhoneNumber,
  isNotValidEmail,
  isNotLetter,
} from './validateHelpers'

export const validateText: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidLength(value)) {
    return 'Pole musi zawierać między 2 - 100 znaków'
  }

  if (isNotLetter(value)) {
    return 'Podane dane są nieprawidłowe'
  }
}

export const validateEmail: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany email jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidEmail(value)) {
    return customErrorMsg
  }
}

export const validateTelNumber: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany numer telefonu jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidPhoneNumber(value)) {
    return customErrorMsg
  }
}

export const validateCheckbox: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => (required && !value && 'required') || undefined
