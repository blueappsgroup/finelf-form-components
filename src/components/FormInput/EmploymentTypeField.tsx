/* eslint-disable @typescript-eslint/camelcase */
import React, { FC } from 'react'
import { Field } from 'formik'

import { SelectFieldProps } from '../../types'
import { BaseSelectField } from './base'
import { validateSelect } from './validate'

const options = {
  fixed: 'umowa o pracę na czas nieokreślony',
  temporary: 'umowa o pracę na czas określony',
  self_employed: 'samozatrudnienie',
  fixed_partial: 'wdowa/wdowiec',
  government: 'stanowisko rządowe',
  unemployed: 'bezrobotny',
  pension: 'emeryt/rencista',
  agriculture: 'gospodarstwo rolne',
  student: 'student',
  housewife: 'gospodyni domowa',
  other: 'inne',
  service_contract: 'umowa zlecenie / umowa o dzieło',
}

const EmploymentTypeField: FC<SelectFieldProps> = ({
  id,
  label,
  name,
  required = false,
  validate,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
    validate={
      (validate && validate(required)) || validateSelect(required, options)
    }
  />
)

EmploymentTypeField.defaultProps = {
  name: 'employment_type',
}

export default EmploymentTypeField
