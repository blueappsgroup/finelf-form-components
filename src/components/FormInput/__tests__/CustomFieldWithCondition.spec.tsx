import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import BaseField from '../base'
import CustomFieldWithCondition from '../CustomFieldWithCondition'
import EmploymentTypeField from '../EmploymentTypeField'
import { validateText } from '../validate'

describe('<CustomFieldWithCondition />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  /* eslint-disable @typescript-eslint/camelcase */

  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    fieldId = 'customFieldWithCondition',
    fieldName = 'customFieldWithCondition',
    required = false,
    showError = false,
    validate = validateText,
    requiredCondition,
    visibleCondition = { employment_type: ['self_employed'] },
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <EmploymentTypeField />
        <CustomFieldWithCondition
          id={fieldId}
          component={BaseField}
          required={required}
          showError={showError}
          name={fieldName}
          validate={validate}
          requiredCondition={requiredCondition}
          visibleCondition={visibleCondition}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const field = container.querySelector(`[name="${fieldName}"]`)
    const employmentTypeField = container.querySelector(
      '[name="employment_type"]'
    )

    return {
      field,
      employmentTypeField,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { field } = setupWrapper({
      visibleCondition: null,
    })

    expect(field).toBeTruthy()
    expect(field).toMatchSnapshot()
  })

  it('matches snapshot without visibleCondition met', () => {
    const { field } = setupWrapper({})

    expect(field).toBeFalsy()
    expect(field).toMatchSnapshot()
  })

  it('matches snapshot with visibleCondition met', async () => {
    const wrapper = setupWrapper({})
    const { employmentTypeField, container } = wrapper

    await act(async () => {
      fireEvent.change(employmentTypeField, {
        target: { value: 'self_employed' },
      })
    })

    await act(async () => {
      fireEvent.blur(employmentTypeField, {
        target: { value: 'self_employed' },
      })
    })

    expect(
      container.querySelector('[name="customFieldWithCondition"]')
    ).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot without requiredCondition met', () => {
    const { field } = setupWrapper({
      requiredCondition: { employment_type: ['self_employed'] },
    })

    expect(field).toBeFalsy()
    expect(field).toMatchSnapshot()
  })

  it('matches snapshot with requiredCondition met', async () => {
    const wrapper = setupWrapper({
      requiredCondition: { employment_type: ['self_employed'] },
    })
    const { employmentTypeField, container } = wrapper

    await act(async () => {
      fireEvent.change(employmentTypeField, {
        target: { value: 'self_employed' },
      })
    })

    await act(async () => {
      fireEvent.blur(employmentTypeField, {
        target: { value: 'self_employed' },
      })
    })

    expect(
      container.querySelector('[name="customFieldWithCondition"]')
    ).toBeTruthy()
    expect(container).toMatchSnapshot()
  })
})
