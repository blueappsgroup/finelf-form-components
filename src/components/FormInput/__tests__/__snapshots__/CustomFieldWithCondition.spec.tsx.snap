// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomFieldWithCondition /> matches snapshot 1`] = `
<input
  class="sc-fnVYJo csxChD"
  id="customFieldWithCondition"
  name="customFieldWithCondition"
  value=""
/>
`;

exports[`<CustomFieldWithCondition /> matches snapshot with requiredCondition met 1`] = `
<div>
  <form
    action="#"
    class="sc-fujyUd bFqltC"
    id="testForm"
  >
    <input
      class="sc-iCoHVE gEjxKq"
      form="[object Object]"
    />
    <div
      class="sc-iqAbSa VlpJw"
    >
      <div
        class="sc-iemXMA clxLOp"
      >
        <select
          class="sc-dIvqjp cnqptG"
          name="employment_type"
          type="text"
        >
          <option
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="fixed"
          >
            umowa o pracę na czas nieokreślony
          </option>
          <option
            value="temporary"
          >
            umowa o pracę na czas określony
          </option>
          <option
            value="self_employed"
          >
            samozatrudnienie
          </option>
          <option
            value="fixed_partial"
          >
            umowa o pracę na czas nieokreślony (niepełny etat)
          </option>
          <option
            value="government"
          >
            stanowisko rządowe
          </option>
          <option
            value="unemployed"
          >
            bezrobotny
          </option>
          <option
            value="pension"
          >
            emeryt/rencista
          </option>
          <option
            value="agriculture"
          >
            gospodarstwo rolne
          </option>
          <option
            value="student"
          >
            student
          </option>
          <option
            value="housewife"
          >
            gospodyni domowa
          </option>
          <option
            value="other"
          >
            inne
          </option>
          <option
            value="service_contract"
          >
            umowa zlecenie / umowa o dzieło
          </option>
        </select>
        <span
          class="sc-hHEjAm gimkwY"
        />
      </div>
    </div>
    <div
      class="sc-iqAbSa VlpJw"
    >
      <div
        class="sc-lmgQde iscgnI"
      >
        <input
          class="sc-fnVYJo csxChD"
          id="customFieldWithCondition"
          name="customFieldWithCondition"
          required=""
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`<CustomFieldWithCondition /> matches snapshot with visibleCondition met 1`] = `
<div>
  <form
    action="#"
    class="sc-fujyUd bFqltC"
    id="testForm"
  >
    <input
      class="sc-iCoHVE gEjxKq"
      form="[object Object]"
    />
    <div
      class="sc-iqAbSa VlpJw"
    >
      <div
        class="sc-iemXMA clxLOp"
      >
        <select
          class="sc-dIvqjp cnqptG"
          name="employment_type"
          type="text"
        >
          <option
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="fixed"
          >
            umowa o pracę na czas nieokreślony
          </option>
          <option
            value="temporary"
          >
            umowa o pracę na czas określony
          </option>
          <option
            value="self_employed"
          >
            samozatrudnienie
          </option>
          <option
            value="fixed_partial"
          >
            umowa o pracę na czas nieokreślony (niepełny etat)
          </option>
          <option
            value="government"
          >
            stanowisko rządowe
          </option>
          <option
            value="unemployed"
          >
            bezrobotny
          </option>
          <option
            value="pension"
          >
            emeryt/rencista
          </option>
          <option
            value="agriculture"
          >
            gospodarstwo rolne
          </option>
          <option
            value="student"
          >
            student
          </option>
          <option
            value="housewife"
          >
            gospodyni domowa
          </option>
          <option
            value="other"
          >
            inne
          </option>
          <option
            value="service_contract"
          >
            umowa zlecenie / umowa o dzieło
          </option>
        </select>
        <span
          class="sc-hHEjAm gimkwY"
        />
      </div>
    </div>
    <div
      class="sc-iqAbSa VlpJw"
    >
      <div
        class="sc-lmgQde iscgnI"
      >
        <input
          class="sc-fnVYJo csxChD"
          id="customFieldWithCondition"
          name="customFieldWithCondition"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`<CustomFieldWithCondition /> matches snapshot without requiredCondition met 1`] = `null`;

exports[`<CustomFieldWithCondition /> matches snapshot without visibleCondition met 1`] = `null`;
